salt/utils/data.py

```diff
@@ -101,7 +103,8 @@ def decode_dict(data, encoding=None, errors='strict', preserve_dict_class=False,
             try:
                 key = salt.utils.stringutils.to_unicode(key, encoding, errors)
             except TypeError:
-                pass
+                if isinstance(key, collections.Iterable):
+                    key = decode_list(list(key), encoding, errors, preserve_dict_class, preserve_tuples)
```

Comment:
This bit here might be questionable, since generators are hashable, while lists are not. So, in the event that someone had a dictionary with keys that were generators, and they called either this (or its encode counterpart below), it would cause an exception. Thoughts? Perhaps automagic handling of generators should be toggleable via an argument to the function? If so, I would think we'd still want the default behavior be to turn this on by default.